### PR TITLE
[doc] Add ANTLR documentation

### PR DESCRIPTION
**Pull request by matifraga**
**PR Description:**
We move the "adding_new_language" document and create 2 separate versions, one for JAVACC grammar based languages  and another one for ANTLR grammar based languages. We kept the same steps as JAVACC and make some explanation over the steps we provide for free.

### CODE REVIEW

**Comment by matifraga** [comment]
@adangel I'm not sure if we should change the permalink for both sections to:
`pmd_devdocs_major_adding_new_javacc_language` and `pmd_devdocs_major_adding_new_antlr_language`

File: docs/pages/pmd/devdocs/major_contributions/adding_a_new_antlr_based_language.md [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+## 1.  Start with a new sub-module.
+*   See pmd-swift for examples.
+
+## 2.  Implement an AST parser for your language
+*	ANTLR gives you this for free.
+
+## 3.  Create AST node classes
+*	We provide an [AntlrBaseNode](https://github.com/pmd/pmd/blob/master/pmd-core/src/main/java/net/sourceforge/pmd/lang/ast/AntlrBaseNode.java). 
+*	We override ANTLR auto-generated code to provide this for free, you need to add an ANT script similar to the [swift scrip](https://github.com/pmd/pmd/blob/master/pmd-swift/src/main/ant/antlr4.xml).
```
Comment (jsotuyod) [inline docs/pages/pmd/devdocs/major_contributions/adding_a_new_antlr_based_language.md]:
```suggestion
*	We override ANTLR auto-generated code to provide this for free, you need to add an ANT script similar to the [swift script](https://github.com/pmd/pmd/blob/master/pmd-swift/src/main/ant/antlr4.xml).
```

File: docs/pages/pmd/devdocs/major_contributions/adding_a_new_antlr_based_language.md [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+
+## 1.  Start with a new sub-module.
+*   See pmd-swift for examples.
+
+## 2.  Implement an AST parser for your language
+*	ANTLR gives you this for free.
+
+## 3.  Create AST node classes
+*	We provide an [AntlrBaseNode](https://github.com/pmd/pmd/blob/master/pmd-core/src/main/java/net/sourceforge/pmd/lang/ast/AntlrBaseNode.java).
```
Comment (jsotuyod) [inline docs/pages/pmd/devdocs/major_contributions/adding_a_new_antlr_based_language.md]:
```suggestion
*	We provide an [`AntlrBaseNode`](https://github.com/pmd/pmd/blob/master/pmd-core/src/main/java/net/sourceforge/pmd/lang/ast/AntlrBaseNode.java). 
```

File: docs/pages/pmd/devdocs/major_contributions/adding_a_new_antlr_based_language.md [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+## 1.  Start with a new sub-module.
+*   See pmd-swift for examples.
+
+## 2.  Implement an AST parser for your language
+*	ANTLR gives you this for free.
+
+## 3.  Create AST node classes
+*	We provide an [AntlrBaseNode](https://github.com/pmd/pmd/blob/master/pmd-core/src/main/java/net/sourceforge/pmd/lang/ast/AntlrBaseNode.java). 
+*	We override ANTLR auto-generated code to provide this for free, you need to add an ANT script similar to the [swift scrip](https://github.com/pmd/pmd/blob/master/pmd-swift/src/main/ant/antlr4.xml).
+*	You can extend AntlrBaseNode and override any method that you require, but on most cases you won't need to do anything.
```
Comment (jsotuyod) [inline docs/pages/pmd/devdocs/major_contributions/adding_a_new_antlr_based_language.md]:
```suggestion
*	You can extend `AntlrBaseNode` and override any method that you require, but on most cases you won't need to do anything.
```

File: docs/pages/pmd/devdocs/major_contributions/adding_a_new_antlr_based_language.md [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+
+## 4.  Compile your parser 
+*	We override ANTLR auto-generated code to provide this for free, similar to the step before, you will need to use the ANT script.
+*	You should review the [swift pom](https://github.com/pmd/pmd/blob/master/pmd-swift/pom.xml). Don't forget to enable visitor generation property. 
+
+## 5.  Create a TokenManager
+*   We provide a default implementation using [AntlrTokenManager](https://github.com/pmd/pmd/blob/master/pmd-core/src/main/java/net/sourceforge/pmd/lang/antlr/AntlrTokenManager.java) that uses an [AntlrTokenizer](https://github.com/pmd/pmd/blob/master/pmd-core/src/main/java/net/sourceforge/pmd/cpd/AntlrTokenizer.java).
```
Comment (jsotuyod) [inline docs/pages/pmd/devdocs/major_contributions/adding_a_new_antlr_based_language.md]:
```suggestion
*   We provide a default implementation using [`AntlrTokenManager`](https://github.com/pmd/pmd/blob/master/pmd-core/src/main/java/net/sourceforge/pmd/lang/antlr/AntlrTokenManager.java) that uses an [`AntlrTokenizer`](https://github.com/pmd/pmd/blob/master/pmd-core/src/main/java/net/sourceforge/pmd/cpd/AntlrTokenizer.java). 
```

File: docs/pages/pmd/devdocs/major_contributions/adding_a_new_antlr_based_language.md [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+*	We override ANTLR auto-generated code to provide this for free, you need to add an ANT script similar to the [swift scrip](https://github.com/pmd/pmd/blob/master/pmd-swift/src/main/ant/antlr4.xml).
+*	You can extend AntlrBaseNode and override any method that you require, but on most cases you won't need to do anything.
+
+## 4.  Compile your parser 
+*	We override ANTLR auto-generated code to provide this for free, similar to the step before, you will need to use the ANT script.
+*	You should review the [swift pom](https://github.com/pmd/pmd/blob/master/pmd-swift/pom.xml). Don't forget to enable visitor generation property. 
+
+## 5.  Create a TokenManager
+*   We provide a default implementation using [AntlrTokenManager](https://github.com/pmd/pmd/blob/master/pmd-core/src/main/java/net/sourceforge/pmd/lang/antlr/AntlrTokenManager.java) that uses an [AntlrTokenizer](https://github.com/pmd/pmd/blob/master/pmd-core/src/main/java/net/sourceforge/pmd/cpd/AntlrTokenizer.java). 
+*	You must create your own AntlrTokenizer such as we do with [SwiftTokenizer](https://github.com/pmd/pmd/blob/master/pmd-swift/src/main/java/net/sourceforge/pmd/cpd/SwiftTokenizer.java).
```
Comment (jsotuyod) [inline docs/pages/pmd/devdocs/major_contributions/adding_a_new_antlr_based_language.md]:
```suggestion
*	You must create your own `AntlrTokenizer` such as we do with [`SwiftTokenizer`](https://github.com/pmd/pmd/blob/master/pmd-swift/src/main/java/net/sourceforge/pmd/cpd/SwiftTokenizer.java).
```

File: docs/pages/pmd/devdocs/major_contributions/adding_a_new_antlr_based_language.md [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+*	If you wish to filter specific tokens you can create your own implementation of [BaseTokenFilter](https://github.com/pmd/pmd/blob/master/pmd-core/src/main/java/net/sourceforge/pmd/cpd/token/internal/BaseTokenFilter.java) as we did with [SwiftTokenFilter](https://github.com/pmd/pmd/blob/master/pmd-core/src/main/java/net/sourceforge/pmd/cpd/token/AntlrTokenFilter.java).
```
Comment (jsotuyod) [inline docs/pages/pmd/devdocs/major_contributions/adding_a_new_antlr_based_language.md]:
```suggestion
*	If you wish to filter specific tokens you can create your own implementation of [`BaseTokenFilter`](https://github.com/pmd/pmd/blob/master/pmd-core/src/main/java/net/sourceforge/pmd/cpd/token/internal/BaseTokenFilter.java) as we did with [`SwiftTokenFilter`](https://github.com/pmd/pmd/blob/master/pmd-core/src/main/java/net/sourceforge/pmd/cpd/token/AntlrTokenFilter.java).
```

Comment (jsotuyod) [inline docs/pages/pmd/devdocs/major_contributions/adding_a_new_antlr_based_language.md]: if we don't want to, what do we do?

File: docs/pages/pmd/devdocs/major_contributions/adding_a_new_antlr_based_language.md [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+
+## 6.  Create a PMD parser “adapter”
+*   We provide a [BaseParser](https://github.com/pmd/pmd/blob/master/pmd-core/src/main/java/net/sourceforge/pmd/lang/antlr/AntlrBaseParser.java) implementation that you need to extend to create your own adapter as we do with [SwiftParserAdapter](https://github.com/pmd/pmd/blob/master/pmd-swift/src/main/java/net/sourceforge/pmd/lang/swift/SwiftParserAdapter.java).
```
Comment (jsotuyod) [inline docs/pages/pmd/devdocs/major_contributions/adding_a_new_antlr_based_language.md]:
```suggestion
*   We provide a [`BaseParser`](https://github.com/pmd/pmd/blob/master/pmd-core/src/main/java/net/sourceforge/pmd/lang/antlr/AntlrBaseParser.java) implementation that you need to extend to create your own adapter as we do with [`SwiftParserAdapter`](https://github.com/pmd/pmd/blob/master/pmd-swift/src/main/java/net/sourceforge/pmd/lang/swift/SwiftParserAdapter.java).
```

File: docs/pages/pmd/devdocs/major_contributions/adding_a_new_antlr_based_language.md [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+
+## 7.  Create a rule violation factory
+*	We provide a [AntlrRuleViolationFactory](https://github.com/pmd/pmd/blob/master/pmd-core/src/main/java/net/sourceforge/pmd/lang/antlr/AntlrRuleViolationFactory.java) as base implementation, you can use that for most scenarios.
```
Comment (jsotuyod) [inline docs/pages/pmd/devdocs/major_contributions/adding_a_new_antlr_based_language.md]:
```suggestion
*	We provide a [`AntlrRuleViolationFactory`](https://github.com/pmd/pmd/blob/master/pmd-core/src/main/java/net/sourceforge/pmd/lang/antlr/AntlrRuleViolationFactory.java) as base implementation, you can use that for most scenarios.
```

File: docs/pages/pmd/devdocs/major_contributions/adding_a_new_antlr_based_language.md [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+*   The purpose of this class is to create a rule violation instance for your handler (spoiler).
+
+## 8.  Create a version handler
+*   Now you need to create your version handler, as we did with [SwiftHandler](https://github.com/pmd/pmd/blob/master/pmd-swift/src/main/java/net/sourceforge/pmd/lang/swift/SwiftHandler.java).
```
Comment (jsotuyod) [inline docs/pages/pmd/devdocs/major_contributions/adding_a_new_antlr_based_language.md]:
```suggestion
*   Now you need to create your version handler, as we did with [`SwiftHandler`](https://github.com/pmd/pmd/blob/master/pmd-swift/src/main/java/net/sourceforge/pmd/lang/swift/SwiftHandler.java).
```

File: docs/pages/pmd/devdocs/major_contributions/adding_a_new_antlr_based_language.md [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+*   This class is sort of a gateway between PMD and all parsing logic specific to your language. It has 2 purposes:
+    *   `getRuleViolationFactory` method returns an instance of your rule violation factory *(see step #7)*
+    *   `getParser` returns an instance of your parser adapter *(see step #6)*
+
+## 9.  Create a parser visitor adapter
+*	We provide an [AbstractAntlrVisitor](https://github.com/pmd/pmd/blob/master/pmd-core/src/main/java/net/sourceforge/pmd/lang/antlr/AbstractAntlrVisitor.java) as default implementation, to be able to use this you should also add it to the ANT script we talked about on step #3
```
Comment (jsotuyod) [inline docs/pages/pmd/devdocs/major_contributions/adding_a_new_antlr_based_language.md]:
```suggestion
*	We provide an [`AbstractAntlrVisitor`](https://github.com/pmd/pmd/blob/master/pmd-core/src/main/java/net/sourceforge/pmd/lang/antlr/AbstractAntlrVisitor.java) as default implementation, to be able to use this you should also add it to the ANT script we talked about on step #3
```

File: docs/pages/pmd/devdocs/major_contributions/adding_a_new_antlr_based_language.md [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+*   The purpose of this class is to serve as a pass-through `visitor` implementation, which, for all AST types in your language, just executes visit on the base AST type
+
+## 10. Create a rule chain visitor
+*   We provide an [AntlrRuleChainVisitor](https://github.com/pmd/pmd/blob/master/pmd-core/src/main/java/net/sourceforge/pmd/lang/antlr/AntlrRuleChainVisitor.java), you can use that for most scenarios.
```
Comment (jsotuyod) [inline docs/pages/pmd/devdocs/major_contributions/adding_a_new_antlr_based_language.md]:
```suggestion
*   We provide an [`AntlrRuleChainVisitor`](https://github.com/pmd/pmd/blob/master/pmd-core/src/main/java/net/sourceforge/pmd/lang/antlr/AntlrRuleChainVisitor.java), you can use that for most scenarios.
```

File: docs/pages/pmd/devdocs/major_contributions/adding_a_new_antlr_based_language.md [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+
+## 11. Make PMD recognize your language
+*   Create your own subclass of `net.sourceforge.pmd.lang.BaseLanguageModule`, see Swift as an example.
+*   You’ll need to refer the rule chain visitor created in step #10.
+*   Add for each version of your language a call to `addVersion` in your language module’s constructor.
+*   Create the service registration via the text file `src/main/resources/META-INF/services/net.sourceforge.pmd.lang.Language`. Add your fully qualified class name as a single line into it.
+
+## 12. Create an abstract rule class for the language
+*	You need to create your own `AbstractRule`, our AbstractAntlrVisitor implements this and makes the connection with ANTLR via our ANT script (see step #3).
+*	You will have an auto-generated XBaseVisitor class (similar to SwiftBaseVisitor) that you will have to extend as we did with [AbstractSwiftRule](https://github.com/pmd/pmd/blob/master/pmd-swift/src/main/java/net/sourceforge/pmd/lang/swift/AbstractSwiftRule.java).
```
Comment (jsotuyod) [inline docs/pages/pmd/devdocs/major_contributions/adding_a_new_antlr_based_language.md]:
```suggestion
*	You will have an auto-generated `XBaseVisitor` class (similar to `SwiftBaseVisitor`) that you will have to extend as we did with [`AbstractSwiftRule`](https://github.com/pmd/pmd/blob/master/pmd-swift/src/main/java/net/sourceforge/pmd/lang/swift/AbstractSwiftRule.java).
```

File: docs/pages/pmd/devdocs/major_contributions/adding_a_new_antlr_based_language.md [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+    *   `visit` method should evaluate what kind of rule is being applied, and execute appropriate logic. Usually it will just check if the rule is a "parser visitor" kind of rule specific to your language, then execute the visitor. If it’s an XPath rule, then we just need to execute evaluate on that.
+
+## 11. Make PMD recognize your language
+*   Create your own subclass of `net.sourceforge.pmd.lang.BaseLanguageModule`, see Swift as an example.
+*   You’ll need to refer the rule chain visitor created in step #10.
+*   Add for each version of your language a call to `addVersion` in your language module’s constructor.
+*   Create the service registration via the text file `src/main/resources/META-INF/services/net.sourceforge.pmd.lang.Language`. Add your fully qualified class name as a single line into it.
+
+## 12. Create an abstract rule class for the language
+*	You need to create your own `AbstractRule`, our AbstractAntlrVisitor implements this and makes the connection with ANTLR via our ANT script (see step #3).
```
Comment (jsotuyod) [inline docs/pages/pmd/devdocs/major_contributions/adding_a_new_antlr_based_language.md]:
```suggestion
*	You need to create your own `AbstractRule`, our `AbstractAntlrVisitor` implements this and makes the connection with ANTLR via our ANT script (see step #3).
```

File: docs/pages/pmd/devdocs/major_contributions/adding_a_new_antlr_based_language.md (hunk 6 above)
Comment (matifraga) [reply docs/pages/pmd/devdocs/major_contributions/adding_a_new_antlr_based_language.md]: I've just updated this scenario, pointing to AntlrTokenFilter, the default implementation we provide

**Comment by matifraga** [comment]
🥳🥳 @jsotuyod @tomidelucca @lsoncini

**Pull request merged**